Code/Tools/AssetProcessor/native/FileWatcher/FileWatcher.cpp

```diff
@@ -6,154 +6,119 @@
-    Q_ASSERT(invoked);
-}
+    if (sizeA <= sizeB)
+    {
+        return false;
+    }
 
-void FolderRootWatch::ProcessModifyFileEvent(const QString& file)
-{
-    FileChangeInfo info;
-    info.m_action = FileAction::FileAction_Modified;
-    info.m_filePath = file;
-    const bool invoked = QMetaObject::invokeMethod(m_fileWatcher, "AnyFileChange", Qt::QueuedConnection, Q_ARG(FileChangeInfo, info));
-    Q_ASSERT(invoked);
+    QChar slash1 = QChar('\\');
+    QChar slash2 = QChar('/');
+    int posA = 0;
+
+    // A is going to be the longer one, so use B:
+    for (int idx = 0; idx < sizeB; ++idx)
+    {
+        QChar charAtA = folderA.at(posA);
+        QChar charAtB = folderB.at(idx);
+
+        if ((charAtB == slash1) || (charAtB == slash2))
+        {
+            if ((charAtA != slash1) && (charAtA != slash2))
+            {
+                return false;
+            }
+            ++posA;
+        }
+        else
+        {
+            if (charAtA.toLower() != charAtB.toLower())
+            {
+                return false;
+            }
+            ++posA;
+        }
+    }
+    return true;
 }
 
 //////////////////////////////////////////////////////////////////////////
 /// FileWatcher
 FileWatcher::FileWatcher()
-    : m_nextHandle(0)
+    : m_platformImpl(AZStd::make_unique<PlatformImplementation>())
 {
-    qRegisterMetaType<FileChangeInfo>("FileChangeInfo");
+    auto makeFilter = [this](auto signal)
+    {
+        return [this, signal](QString path)
+        {
+            const auto foundWatchRoot = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [path](const WatchRoot& watchRoot)
+            {
+                return Filter(path, watchRoot);
+            });
+            if (foundWatchRoot == end(m_folderWatchRoots))
+            {
+                return;
+            }
+            AZStd::invoke(signal, this, path);
+        };
+    };
+
+    connect(this, &FileWatcher::rawFileAdded, makeFilter(&FileWatcher::fileAdded));
+    connect(this, &FileWatcher::rawFileRemoved, makeFilter(&FileWatcher::fileRemoved));
+    connect(this, &FileWatcher::rawFileModified, makeFilter(&FileWatcher::fileModified));
 }
 
 FileWatcher::~FileWatcher()
 {
+    disconnect();
+    StopWatching();
 }
 
-int FileWatcher::AddFolderWatch(FolderWatchBase* pFolderWatch)
+void FileWatcher::AddFolderWatch(QString directory, bool recursive)
 {
-    if (!pFolderWatch)
-    {
-        return -1;
-    }
-
-    FolderRootWatch* pFolderRootWatch = nullptr;
-
-    //see if this a sub folder of an already watched root
-    for (auto rootsIter = m_folderWatchRoots.begin(); !pFolderRootWatch && rootsIter != m_folderWatchRoots.end(); ++rootsIter)
+    // Search for an already monitored root that is a parent of `directory`,
+    // that is already watching subdirectories recursively
+    const auto found = AZStd::find_if(begin(m_folderWatchRoots), end(m_folderWatchRoots), [directory](const WatchRoot& root)
```

Comment:
This loop probably should be turned on its side. Instead of scanning all watched roots, we could just iterate the parent paths of `directory` and lookup each one.

Reply:
I'm not following. `m_folderWatchRoots` is just a `vector`.
> iterate the parent paths of `directory` and lookup each one.

In what container are the parent paths of `directory` being looked up?